### PR TITLE
bugfix: InsertExecutor afterImage

### LINKED ISSUE #1500: discuss InsertExecutor#afterImage
@slievrly @zjinlei 
primary key appears case:
1. 
```
create table test(id bigint AUTO_INCREMENT primary key, name varchar(20));

insert into test values(null, ?) 
--> getPkValuesByAuto (right)

insert into test(id, name) values(null, ?) 
--> getPkValuesByColumn (error)
expected: pkValues size is 1

insert into test(name) values(?)
--> getPkValuesByAuto (right)

insert into test values(null, ?), (null, ?)
--> getPkValuesByAuto
expected: pkValues size is 2

insert into test(id, name) values(null, ?), (null, ?)
--> getPkValuesByColumn
expected: pkValues size is 2

insert into test(name) values(?), (?)
--> getPkValuesByAuto (right)
```
I don't know if there is any problem with understanding.

### PR DESCRIPTION
**Pull request by jsbxyyx**
### Ⅰ. Describe what this PR did
fix InsertExecutor afterImage

### Ⅱ. Does this pull request fix one issue?

fixes #1500 #1430 #1516 

### Ⅳ. Describe how to verify it

mysql
create table test(
id int primary key auto_increment, 
name varchar(45), 
name2 varchar(45)
)

1. containsPK (preparedstatement)  pk is ?
insert into test(id, name, name2) values (?, ?, ?)
2. containsPK (preparedstatement) pk is null
insert into test(id, name, name2) values(null, ?, ?)
3. containsPK (statement) pk is null
insert into test(id, name, name2) values(null, '11', '111')
4. containsPK (preparedstatement) pk is null or ? is null
insert into test(id, name, name2) values(null, ?, ?), (?, ?, ?)
parameters: ['11', '111', null, '22', '222']
5. not containsPK (preparedstatement) pkvalue is null
insert into test values(null, ?, ?)
6. not containsPK (preparedstatement) pkvalue is ?
insert into test values(?, ?, ?)
7. not containsPK (preparedstatement)
insert into test(name, name2) values(?, ?)
8. not containsPK (statement)
insert into test(name, name3) values('11', '111')

mysql not support
1. insert into test(id, name, name2) values(fun(), ?, ?), (fun(), ?, ?)
2. insert into test(id, name, name2) values(null, '11', '111'), (null, '22', '222')
excludes specify Statement.RETURN_GENERATED_KEYS case
3. insert into test(id, name, name2) values(?, ?, ?), (1, ?, ?)

oracle
create table test(
id int primary key, 
name varchar(45), 
name2 varchar(45)
)
create sequence test_seq;

1. containsPK (preparedstatement) pk is seq
insert into test(id, name, name2) values(test_seq.nextval, ?, ?)
2. not containsPK (preparedstatement)
insert into test values(test_seq.nextval, ?, ?)

oracle not support
1. insert into test(id, name, name2) values(fun(), ?, ?), (fun(), ?, ?)
2. insert into test(id, name, name2) values(test_seq.nextval, '11', '111'), (test_seq.nextval, '22', '222')
3. insert into test(id, name, name2) values(test_seq.nextval, ?, ?), (1, ?, ?)

### CODE REVIEW

File: test/src/test/java/io/seata/core/rpc/netty/v1/ProtocolV1CodecTest.java [hunk 1]
```diff
@@ -92,7 +92,7 @@ public void testAll() {
                 });
             }
 
-            cnt.await(10, TimeUnit.SECONDS);
+            cnt.await();
```
Comment (zjinlei) [inline test/src/test/java/io/seata/core/rpc/netty/v1/ProtocolV1CodecTest.java]: why do it in this pr?

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java [hunk 2]
```diff
@@ -95,25 +95,52 @@ protected boolean containsPK() {
...
+                    if (parameters.length % rowSize == 0 && rowSize != 1) {
+                        int placeholderNum = parameters.length / rowSize;
+                        pkValues = new ArrayList<>(rowSize);
+                        for (int i = 0; i < rowSize; i++) {
+                            List<Object> paramsByIndex = preparedStatementProxy.getParamsByIndex(placeholderNum * i + pkIndex);
+                            pkValues.add(paramsByIndex.get(0));
+                        }
+                        return pkValues;
+                    }
+                    pkValues = preparedStatementProxy.getParamsByIndex(pkIndex);
```
[the remark this answers is not in the document]
Comment (jsbxyyx) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: ok, i update it

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java [hunk 3]
```diff
@@ -95,25 +95,52 @@ protected boolean containsPK() {
...
+                    return pkValues;
+                }
+
+                if (hasPlaceholder) {
+                    ArrayList<Object>[] parameters = preparedStatementProxy.getParameters();
+                    if (parameters.length % rowSize == 0 && rowSize != 1) {
+                        int placeholderNum = parameters.length / rowSize;
+                        pkValues = new ArrayList<>(rowSize);
+                        for (int i = 0; i < rowSize; i++) {
+                            List<Object> paramsByIndex = preparedStatementProxy.getParamsByIndex(placeholderNum * i + pkIndex);
```
[the remark this answers is not in the document]
Comment (jsbxyyx) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: ok, i update it

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java [hunk 4]
```diff
@@ -95,25 +95,52 @@ protected boolean containsPK() {
...
+            List<List<Object>> insertRows = recognizer.getInsertRows();
+            if (insertRows != null && !insertRows.isEmpty()) {
+                int rowSize = insertRows.size();
+                boolean hasNull = false;
+                boolean hasPlaceholder = false;
+                for (List<Object> row : insertRows) {
+                    Object pkValue = row.get(pkIndex);
+                    if (pkValue instanceof Null) {
+                        hasNull = true;
+                        continue;
```
[the remark this answers is not in the document]
Comment (jsbxyyx) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: here is can update to break

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java [hunk 5]
```diff
@@ -95,25 +95,52 @@ protected boolean containsPK() {
...
-                pkIndexs.add(insertColumnsSize * i + firstPkIndex);
-            }
-            if (pkIndexs.size() == 1) {
-                //adapter test case
-                pkValues = preparedStatementProxy.getParamsByIndex(pkIndexs.get(0));
-            } else {
-                pkValues = pkIndexs.stream().map(pkIndex -> paramters[pkIndex].get(0)).collect(Collectors.toList());
+
+            List<List<Object>> insertRows = recognizer.getInsertRows();
+            if (insertRows != null && !insertRows.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (jsbxyyx) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: should never happen, but InsertExecutorTest#testGetPkValuesByColumn_Exception case use

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java [hunk 6]
```diff
@@ -95,25 +95,52 @@ protected boolean containsPK() {
...
+                boolean hasPlaceholder = false;
+                for (List<Object> row : insertRows) {
+                    Object pkValue = row.get(pkIndex);
+                    if (pkValue instanceof Null) {
+                        hasNull = true;
+                        continue;
+                    }
+                    if ("?".equals(pkValue)) {
+                        hasPlaceholder = true;
+                        continue;
```
[the remark this answers is not in the document]
Comment (jsbxyyx) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: here not update to break, can appears Null value

File: test/src/test/java/io/seata/core/rpc/netty/v1/ProtocolV1CodecTest.java (hunk 1 above)
Comment (jsbxyyx) [reply test/src/test/java/io/seata/core/rpc/netty/v1/ProtocolV1CodecTest.java]: the test case can appears not pass

File: rm-datasource/src/test/java/io/seata/rm/datasource/exec/BatchInsertExecutorTest.java [hunk 7]
```diff
@@ -87,52 +88,66 @@ public void testGetPkValuesByColumn() throws SQLException {
...
+        row.add("?");
+        row.add("?");
+        row.add("?");
+        rows.add(row);
+        row = new ArrayList<>();
+        row.add("?");
+        row.add("?");
+        row.add("?");
+        rows.add(row);
+        when(sqlInsertRecognizer.getInsertRows()).thenReturn(rows);
```
[the remark this answers is not in the document]
Comment (jsbxyyx) [reply rm-datasource/src/test/java/io/seata/rm/datasource/exec/BatchInsertExecutorTest.java]: here is 1, I fix it.

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java [hunk 8]
```diff
@@ -95,25 +95,50 @@ protected boolean containsPK() {
...
+                    }
+                }
+                if (hasNull && hasPlaceholder) {
```
Comment (zjinlei) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: Here we don't know pkValues yet, it is possible that the pkValue of the placeholder is also Null.

Comment (jsbxyyx) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: fix done

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java [hunk 9]
```diff
@@ -95,25 +97,55 @@ protected boolean containsPK() {
...
+                        for (Object r : row) {
+                            if (PLACEHOLDER.equals(r)) {
+                                totalPlaceholder += 1;
+                                placeholderNum += 1;
+                            }
+                        }
+                        if (PLACEHOLDER.equals(pkValue)) {
+                            int idx = pkIndex;
+                            if (i != 0) {
+                                idx = totalPlaceholder - (placeholderNum == pkIndex ? 0 : pkIndex);
```
Comment (zjinlei) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: : pkIndex?

Comment (jsbxyyx) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: idx = totalPlaceholder - (placeholderNum == pkIndex ? 0 : placeholderNum  - pkIndex);

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java [hunk 10]
```diff
@@ -166,8 +167,12 @@ protected boolean containsPK() {
         if (b != 0) {
             throw new NotSupportYetException("not support sql [" + recognizer.getOriginalSQL() + "]");
         }
+        // pk auto generated while single insert primary key is expression
+        if (pkValues.size() == 1 && pkValues.get(0) instanceof SqlMethodExpr) {
+            pkValues = getPkValuesByAuto();
+        }
         // pk auto generated while column exists and value is null
-        if (pkValues.size() > 0 && pkValues.get(0) instanceof Null) {
+        else if (pkValues.size() > 0 && pkValues.get(0) instanceof Null) {
```
Comment (zjinlei) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: pkValues.size() > 0 && pkValues.stream().allMatch(pkValue -> pkValue instanceof Null)
easy to understand

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java [hunk 11]
```diff
@@ -209,7 +214,7 @@ protected boolean containsPK() {
     /**
      * check pk values
      * @param pkValues
-     * @return 0 ok 1 exists null value and normal value
+     * @return 0 ok 1 exists null value and non-null value
      */
     private int checkPkValues(List<Object> pkValues) {
```
Comment (zjinlei) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: 1.why not return boolean
2.unsupport SqlMethodExpr size>1

Comment (jsbxyyx) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: fix done

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java [hunk 12]
```diff
@@ -69,7 +73,7 @@ protected TableRecords beforeImage() throws SQLException {
     @Override
     protected TableRecords afterImage(TableRecords beforeImage) throws SQLException {
         //Pk column exists or PK is just auto generated
-        List<Object> pkValues = containsPK() ? getPkValuesByColumn() : getPkValuesByAuto();
+        List<Object> pkValues = containsPK() ? getPkValuesByColumn() : containsColumns() ? getPkValuesByAuto() : getPkValuesByColumn();
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: For readability, it is recommended to add a parenthesis to indicate its priority.

**Comment by jsbxyyx** [comment]
```
mysql
create table test(
id int primary key auto_increment, 
name varchar(45), 
name2 varchar(45)
)

1. containsPK (preparedstatement)  pk is ?
insert into test(id, name, name2) values (?, ?, ?)
2. containsPK (preparedstatement) pk is null
insert into test(id, name, name2) values(null, ?, ?)
3. containsPK (statement) pk is null
insert into test(id, name, name2) values(null, '11', '111')
4. containsPK (preparedstatement) pk is null or ?
insert into test(id, name, name2) values(null, ?, ?), (?, ?, ?)
parameters: ['11', '111', null, '22', '222']
5. not containsPK (preparedstatement) pkvalue is null
insert into test values(null, ?, ?)
6. not containsPK (preparedstatement) pkvalue is ?
insert into test values(?, ?, ?)
7. not containsPK (preparedstatement)
insert into test(name, name2) values(?, ?)
8. not containsPK (statement)
insert into test(name, name3) values('11', '111')

mysql not support
1. insert into test(id, name, name2) values(fun(), ?, ?), (fun(), ?, ?)
2. insert into test(id, name, name2) values(null, '11', '111'), (null, '22', '222')
3. insert into test(id, name, name2) values(?, ?, ?), (1, ?, ?)

oracle
create table test(
id int primary key, 
name varchar(45), 
name2 varchar(45)
)
create sequence test_seq;

1. containsPK (preparedstatement) pk is seq
insert into test(id, name, name2) values(test_seq.nextval, ?, ?)
2. not containsPK (preparedstatement)
insert into test values(test.seq_nextval, ?, ?)

oracle not support
1. insert into test(id, name, name2) values(fun(), ?, ?), (fun(), ?, ?)
2. insert into test(id, name, name2) values(test_seq.nextval, '11', '111'), (test_seq.nextval, '22', '222')
3. insert into test(id, name, name2) values(test_seq.nextval, ?, ?), (1, ?, ?)

```

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java (hunk 12 above)
Comment (jsbxyyx) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: fix done

**Pull request merged**